File: rust/examples/list_references.py
"""Espejo Python de list_references.rs."""
import sys

import gdstk


def main() -> None:
    if len(sys.argv) < 2:
        print("usage: list_references.py <file.gds>", file=sys.stderr)
        sys.exit(2)

    lib = gdstk.read_gds(sys.argv[1])
    for cell in lib.cells:
        if not cell.references:
            continue
        print(f"Cell '{cell.name}':")
        for r in cell.references:
            # r.cell can be a Cell, RawCell, or str (ReferenceType::Name).
            if isinstance(r.cell, str):
                name = r.cell
            else:
                name = r.cell.name
            x, y = r.origin
            print(
                f"  -> '{name}' @ ({x:.2f}, {y:.2f}) "
                f"rot={r.rotation:.3f} mag={r.magnification:.3f} "
                f"xrefl={str(r.x_reflection).lower()}"
            )


if __name__ == "__main__":
    main()

File: rust/examples/list_references.rs
//! Imprime referencias por celda. Test de paridad Fase 4.

use std::env;

use gdstk_rs::Library;

fn main() {
    let mut args = env::args();
    let program = args.next().unwrap_or_else(|| "list_references".into());
    let Some(path) = args.next() else {
        eprintln!("usage: {program} <file.gds>");
        std::process::exit(2);
    };

    let lib = Library::open(&path);
    for cell in lib.cells() {
        if cell.reference_count() == 0 {
            continue;
        }
        println!("Cell '{}':", cell.name());
        for r in cell.references() {
            let p = r.origin();
            println!(
                "  -> '{}' @ ({:.2}, {:.2}) rot={:.3} mag={:.3} xrefl={}",
                r.cell_name(),
                p.x,
                p.y,
                r.rotation(),
                r.magnification(),
                r.x_reflection()
            );
        }
    }
}
